Add mcxe31b edma support

mcuxted opened the pull request:
1. Remove DMA_MCUX_EDMA_V5 configuration.
2. Add dmamux support with edma.
3. Add mcxe31b edma support.
4. Enable MCXE31B dma_loop_transfer case.
Test Result:
`Running TESTSUITE dma_m2m_loop

===================================================================

START - test_tst_dma0_m2m_loop

DMA memory to memory transfer started

Preparing DMA Controller: edma@20c000

Starting the transfer on channel 0 and waiting for 1 second

Each RX buffer should contain the full TX buffer string.

RX data Loop 0

RX data Loop 1

RX data Loop 2

RX data Loop 3

Finished DMA: edma@20c000

 PASS - test_tst_dma0_m2m_loop in 0.275 seconds

===================================================================

START - test_tst_dma0_m2m_loop_repeated_start_stop

DMA memory to memory transfer started

Preparing DMA Controller

Starting the transfer on channel 0 and waiting for 1 second

Each RX buffer should contain the full TX buffer string.

RX data Loop 0

RX data Loop 1

RX data Loop 2

RX data Loop 3

Finished: DMA

 PASS - test_tst_dma0_m2m_loop_repeated_start_stop in 0.273 seconds

===================================================================

START - test_tst_dma0_m2m_loop_suspend_resume

DMA memory to memory transfer started

Preparing DMA Controller: edma@20c000

Starting the transfer on channel 0 and waiting for 1 second

suspended after 0 transfers occurred

resuming after 0 transfers occurred

Resumed transfers

Transfer count 4

Each RX buffer should contain the full TX buffer string.

RX data Loop 0

RX data Loop 1

RX data Loop 2

RX data Loop 3

Finished DMA: edma@20c000

 PASS - test_tst_dma0_m2m_loop_suspend_resume in 0.535 seconds

===================================================================

TESTSUITE dma_m2m_loop succeeded



------ TESTSUITE SUMMARY START ------



SUITE PASS - 100.00% [dma_m2m_loop]: pass = 3, fail = 0, skip = 0, total = 3 duration = 1.083 seconds

 - PASS - [dma_m2m_loop.test_tst_dma0_m2m_loop] duration = 0.275 seconds

 - PASS - [dma_m2m_loop.test_tst_dma0_m2m_loop_repeated_start_stop] duration = 0.273 seconds

 - PASS - [dma_m2m_loop.test_tst_dma0_m2m_loop_suspend_resume] duration = 0.535 seconds



------ TESTSUITE SUMMARY END ------



===================================================================

PROJECT EXECUTION SUCCESSFUL
`

mcuxted committed:
drivers: dma: mcux_edma: Remove DMA_MCUX_EDMA_V5 configuration

The DMA_MCUX_EDMA_V5 configuration option has been removed and replaced
with DMA_MCUX_EDMA_V4, as both versions share the same register layout
and can use the same driver implementation.

Key changes:
- Remove CONFIG_DMA_MCUX_EDMA_V5 Kconfig option
- Replace DMA_MCUX_EDMA_V5 conditionals with DMA_MCUX_EDMA_V4
- Remove DMAx_Type typedef, use DMA_Type directly
- Update EDMA_HW_TCD macros for V4 to use HAL-provided accessor macros
- Add DMA_MCUX_EDMA_DMAMUX Kconfig option to control DMAMUX support
  based on device tree property
- Update device tree binding to add has-dmamux property
- Update HAL driver selection to use DMA_MCUX_EDMA_DMAMUX instead of
  DMA_MCUX_EDMA for DMAMUX component
- Add SOC_SERIES_MCXE31X to DMA_MCUX_TEST_SLOT_START configuration
- Calculate DMA_TCD_ALIGN_SIZE from edma_tcd_t structure size

Signed-off-by: Qiang Zhang <qiang.zhang_6@nxp.com>

mcuxted committed:
dts: arm: nxp: Update EDMA device tree configurations

Update EDMA device tree nodes for NXP MCXE31B platforms to align
with the unified EDMA driver implementation.

Signed-off-by: Qiang Zhang <qiang.zhang_6@nxp.com>

mcuxted committed:
tests: drivers/dma: Add support for frdm_mcxe31b

Add board-specific configuration and overlay files to enable DMA
testing on the FRDM-MCXE31B board.

Signed-off-by: Qiang Zhang <qiang.zhang_6@nxp.com>

zejiang0jason commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -304,7 +309,6 @@ static void edma_configure_dmamux(const struct device *dev, uint32_t channel,
 
 	/* dam_imx_rt_set_channel_priority(dev, channel, config); */
 	DMAMUX_EnableChannel(DEV_DMAMUX_BASE(dev, dmamux_idx), dmamux_channel);
-}
```
commit by mistake?

zejiang0jason commented on drivers/dma/Kconfig.mcux_edma:
```diff
@@ -51,11 +51,12 @@ config DMA_TCD_QUEUE_SIZE
 	depends on (SOC_SERIES_KINETIS_K6X || SOC_SERIES_KINETIS_KE1XF \
-		    || SOC_SERIES_S32K3 || SOC_SERIES_S32ZE || SOC_SERIES_KE1XZ)
+		    || SOC_SERIES_S32K3 || SOC_SERIES_S32ZE || SOC_SERIES_KE1XZ \
+		    || SOC_SERIES_MCXE31X)
```
This E31 change should be placed in a separate E31 patch.

ZhaoxiangJin commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -40,12 +40,6 @@ LOG_MODULE_REGISTER(dma_mcux_edma, CONFIG_DMA_LOG_LEVEL);
 #define HAS_CHANNEL_GAP(n)		DT_INST_NODE_HAS_PROP(n, channel_gap) ||
 #define DMA_MCUX_HAS_CHANNEL_GAP	(DT_INST_FOREACH_STATUS_OKAY(HAS_CHANNEL_GAP) 0)
 
```
+1, it is better to have a detailed description for Driver-IP-SoC mapping

lucien-nxp commented on drivers/dma/Kconfig.mcux_edma:
```diff
@@ -5,42 +5,47 @@
 	depends on $(dt_compat_any_has_prop,$(EDMA_COMPAT),$(REV_PROP),3)
 	help
-	  DMA version 3 driver for MCUX series SoCs.
+	  DMA version 3 driver for MCUX series SoCs with EDMA3 IP.
```
dma3 IP

mcuxted commented:
Please review whether this patch requires any additional updates. Thanks.

hakehuang commented:
to enable dma, all tests/drivers/dma cases shall pass. not only one.

hakehuang commented:
> Hi @hakehuang ,
> Could you help filter edma case to test for all the nxp platforms to avoid some regression issues? Thank you.

will schedule a full regression, and feedback

ZhaoxiangJin commented on dts/bindings/dma/nxp,mcux-edma.yaml:
```diff
@@ -70,9 +76,14 @@ properties:
       - 4
-      - 5
     description: |
-      eDMA IP revision number.
```
please update the 4.4 migration guide to document this change.

mcuxted committed:
dts: nxp: frdm_mcxe31b: Add LPUART2 DMA support

Add pinctrl configuration and DMA channel assignments for LPUART2
on the FRDM-MCXE31B board.

Signed-off-by: Qiang Zhang <qiang.zhang_6@nxp.com>

mcuxted committed:
tests: drivers/uart: Add async API support for frdm_mcxe31b

Add board-specific overlay file to enable UART async API testing on the
FRDM-MCXE31B board.

Signed-off-by: Qiang Zhang <qiang.zhang_6@nxp.com>

mcuxted commented:
> to enable dma, all tests/drivers/dma cases shall pass. not only one.

all dma demo enableed for frdmmcxe31b.

hakehuang commented on tests/drivers/dma/loop_transfer/boards/imx95_evk_mimx9596_m7.overlay:
```diff
@@ -0,0 +1,9 @@
+/*
```
this is not related to this pr, remove it

hakehuang commented on tests/drivers/dma/chan_link_transfer/boards/frdm_mcxe31b.overlay:
```diff
@@ -0,0 +1,10 @@
+/*
```
need update testcase.yaml in `platfrom_allow`

hakehuang commented on tests/drivers/dma/scatter_gather/boards/frdm_mcxe31b.conf:
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2025 NXP
```
need update testcase.yaml for 'platform_allow'

lucien-nxp commented on tests/drivers/dma/chan_link_transfer/boards/frdm_mcxe31b.overlay:
```diff
@@ -0,0 +1,10 @@
+ */
+
+/* Add test label to edma node */
+dma0: &edma {
```
dma0 enablement need to be moved to board level.

lucien-nxp commented on tests/drivers/dma/scatter_gather/boards/frdm_mcxe31b.overlay:
```diff
@@ -0,0 +1,16 @@
+/*
+ * Copyright (c) 2025 NXP
```
The same issue on license.

hakehuang commented on tests/drivers/uart/uart_async_api/boards/frdm_mcxe31b.overlay:
```diff
@@ -0,0 +1,15 @@
+/*
```
you have using loopback mode in lpuart, so you can just change the testcase.yaml as below
```
  drivers.uart.async_api.lpuart.rt_nocache:
    filter: CONFIG_SERIAL_SUPPORT_ASYNC and CONFIG_UART_MCUX_LPUART and CONFIG_CPU_HAS_DCACHE
    harness: ztest
    depends_on: dma
    extra_configs:
      - CONFIG_DCACHE=y
      - CONFIG_NOCACHE_MEMORY=y
      - CONFIG_USERSPACE=n
      - CONFIG_TEST_USERSPACE=n
    extra_args:
      - platform:mimxrt1010_evk/mimxrt1011:"DTC_OVERLAY_FILE=nxp/dut_lpuart4_loopback.overlay"
      - platform:mimxrt1015_evk/mimxrt1015:"DTC_OVERLAY_FILE=nxp/dut_lpuart4_loopback.overlay"
      - platform:mimxrt1020_evk/mimxrt1021:"DTC_OVERLAY_FILE=nxp/dut_lpuart2_loopback.overlay"
      - platform:mimxrt1024_evk/mimxrt1024:"DTC_OVERLAY_FILE=nxp/dut_lpuart2_loopback.overlay"
      - platform:mimxrt1160_evk/mimxrt1166/cm7:"DTC_OVERLAY_FILE=nxp/dut_lpuart2_loopback.overlay"
      - platform:mimxrt1170_evk@A/mimxrt1176/cm7:"DTC_OVERLAY_FILE=nxp/dut_lpuart2_loopback.overlay"
      - platform:mimxrt1170_evk@B/mimxrt1176/cm7:"DTC_OVERLAY_FILE=nxp/dut_lpuart2_loopback.overlay"
      - platform:vmu_rt1170/mimxrt1176/cm7:"DTC_OVERLAY_FILE=nxp/dut_lpuart4_loopback.overlay"
      - platform:mimxrt1180_evk/mimxrt1189/cm33:"DTC_OVERLAY_FILE=nxp/dut_lpuart3_loopback.overlay"
      - platform:mimxrt1050_evk/mimxrt1052/hyperflash:"DTC_OVERLAY_FILE=nxp/dut_lpuart3_loopback.overlay"
      - platform:mimxrt1060_evk/mimxrt1062/hyperflash:"DTC_OVERLAY_FILE=nxp/dut_lpuart3_loopback.overlay"
      - platform:mimxrt1060_evk@A/mimxrt1062/qspi:"DTC_OVERLAY_FILE=nxp/dut_lpuart3_loopback.overlay"
      - platform:mimxrt1060_evk@B/mimxrt1062/qspi:"DTC_OVERLAY_FILE=nxp/dut_lpuart3_loopback.overlay"
      - platform:mimxrt1060_evk@C/mimxrt1062/qspi:"DTC_OVERLAY_FILE=nxp/dut_lpuart3_loopback.overlay"
      - platform:mimxrt1064_evk/mimxrt1064:"DTC_OVERLAY_FILE=nxp/dut_lpuart3_loopback.overlay"
      - platform:frdm_mcxe31b:"DTC_OVERLAY_FILE=nxp/dut_lpuart2_loopback.overlay"
```

hakehuang commented on tests/drivers/uart/uart_async_api/boards/frdm_mcxe31b.overlay:
```diff
@@ -0,0 +1,15 @@
+/*
+ * Copyright 2025 NXP
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
this file can removed

hakehuang commented on boards/nxp/frdm_mcxe31b/frdm_mcxe31b.dts:
```diff
@@ -204,3 +215,6 @@
 };
+
+/* Add lpuart2 label to lpuart_2. */
+lpuart2: &lpuart_2 {};
```
this is not good. why we can not make this consistent with former platform, as they are the same IP?

hakehuang replied:
As software produce, we need keep consistency. unless absolutely necessary. we can add a `_`, or 'v' in the end of the IP name which does not hurt, but it is not necessary we add _ for all.

lucien-nxp replied:
As I mentioned before, we have no right to define the names of any IP instances. These names are all provided by the design team. The actual situation is that we obtain the source data in frdm_mcxe31b, it is all accompanied by `_`.  RM also record the same name as below:
<img width="296" height="471" alt="image" src="https://github.com/user-attachments/assets/43bf89ff-fddb-4cee-a284-6f95502bcbf4" />

hakehuang replied:
> As I mentioned before, we have no right to define the names of any IP instances. These names are all provided by the design team. The actual situation is that we obtain the source data in frdm_mcxe31b, it is all accompanied by `_`. RM also record the same name as below:

I believe there is no mis-leading with lPUART2 -> LPUART_2, as we have using this paradigm for years. and the only change is about the generate script do a simple processing for IP name end with digital by adding with a `_`, but for legacy ones, the scripts shall keep consistent.

hakehuang replied:
I would suggest to use a default domain and others can use <domain>_<ipname>_<instance>

mcuxted commented:
hi @hakehuang , This patch has been stuck in the lable naming process for too long. Can we test the global demo first to see if there are any other issues with this path?

CkovMk commented:
This PR changes the meaning of "nxp,version" to HAL driver version. All instances of `nxp,mcux-edma` must have the same HAL driver version in a project. This definition has nothing to do with actual hardaware (aka register layout). As I mentioned during off-line meet-up, it's weird to encode this information into device tree. In my view, this kind of difference should be handled in SoC level, not IP level.

As this goes into release note, I think we'd better be cautious.

I purpose following solutions:

Solution 1:
- Create multiple compatibles according to HAL driver version being used.

Solution 2:
- Provide Kconfig options in dma_mcux_edma driver and select HAL driver version in SoC defconfig.

Personally I prefer solution 2. The impact is minimum (we are already using such Kconfigs, just set the value in SoC defconfig instead of via device tree) and it keeps the compatible str of the same IP consistent.

mcuxted commented:
> This PR changes the meaning of "nxp,version" to HAL driver version. All instances of `nxp,mcux-edma` must have the same HAL driver version in a project. This definition has nothing to do with actual hardaware (aka register layout). As I mentioned during off-line meet-up, it's weird to encode this information into device tree. In my view, this kind of difference should be handled in SoC level, not IP level.
> 
> As this goes into release note, I think we'd better be cautious.
> 
> I purpose following solutions:
> 
> Solution 1:
> 
> * Create multiple compatibles according to HAL driver version being used.
> 
> Solution 2:
> 
> * Provide Kconfig options in dma_mcux_edma driver and select HAL driver version in SoC defconfig.
> 
> Personally I prefer solution 2. The impact is minimum (we are already using such Kconfigs, just set the value in SoC defconfig instead of via device tree) and it keeps the compatible str of the same IP consistent.

To preserve the functionality of the NXP Veriton, the selection method can be optimized to choose the corresponding HAL driver based on the EDMA IP version.

lucien-nxp commented:
> It is strange to specify software driver versions through DT node, this may violates the design philosophy of what DT is meant for. We should use Kconfig to handle this kind of software-level difference.

Firstly, I believe this topic can be discussed separately in a dedicated channel, without hindering this PR (as the PR merely follows the current usage method).
Secondly, Declan do this [modification ](https://github.com/zephyrproject-rtos/zephyr/pull/79872)14 months ago, and inspired from a linux DT property from st called st,version on their dma. Whether such information is described in DTS or kconfig, the essence is to facilitate the call at the HAL layer. However, it will cause significant compatibility issues for user upgrades. Please consider carefully before taking any actual actions.

mcuxted committed:
doc: migration-guide-4.4: Document DMA_MCUX_EDMA_V5 removal

Add migration guide entry for the removal of CONFIG_DMA_MCUX_EDMA_V5
configuration option.

Signed-off-by: Qiang Zhang <qiang.zhang_6@nxp.com>

zejiang0jason commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -229,6 +229,13 @@ Counter
+DMA
+===
+
+* Removed ``DMA_MCUX_EDMA_V5`` (:github:`100341`). This macro previously distinguished between
```
use ":kconfig:option:`CONFIG_DMA_MCUX_EDMA_V5`", it is a kconfig option.
